test/test_selectors.py

```diff
@@ -0,0 +1,664 @@
+        s2 = self.make_selector()
+        rd, wr = self.make_socketpair()
+        key1 = s1.register(rd, selectors.EVENT_READ)
+        key2 = s2.register(rd, selectors.EVENT_READ)
+
+        wr.send(b'x')
+        time.sleep(0.01)  # Wait for the write to flush.
+
+        self.assertEqual([(key1, selectors.EVENT_READ)], s1.select(timeout=0.001))
+        self.assertEqual([(key2, selectors.EVENT_READ)], s2.select(timeout=0.001))
+
+    def test_select_no_event_types(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        self.assertRaises(ValueError, s.register, rd, 0)
+
+    def test_select_many_events(self):
+        s = self.make_selector()
+        readers = []
+        writers = []
+        for _ in range(32):
+            rd, wr = self.make_socketpair()
+            readers.append(rd)
+            writers.append(wr)
+            s.register(rd, selectors.EVENT_READ)
+
+        self.assertEqual(0, len(s.select(0.001)))
+
+        # Write a byte to each end.
+        for wr in writers:
+            wr.send(b'x')
+
+        # Give time to flush the writes.
+        time.sleep(0.01)
+
+        ready = s.select(0.001)
+        self.assertEqual(32, len(ready))
+        for key, events in ready:
+            self.assertEqual(selectors.EVENT_READ, events)
+            self.assertIn(key.fileobj, readers)
+
+        # Now read the byte from each endpoint.
+        for rd in readers:
+            data = rd.recv(1)
+            self.assertEqual(b'x', data)
+
+        self.assertEqual(0, len(s.select(0.001)))
+
+    def test_select_timeout_none(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(wr, selectors.EVENT_WRITE)
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(1, len(s.select(timeout=None)))
+
+    def test_select_timeout_ready(self):
+        s, rd, wr = self.standard_setup()
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(1, len(s.select(timeout=0)))
+            self.assertEqual(1, len(s.select(timeout=-1)))
+            self.assertEqual(1, len(s.select(timeout=0.001)))
+
+    def test_select_timeout_not_ready(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=0)))
+
+        with self.assertTakesTime(lower=SHORT_SELECT, upper=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=SHORT_SELECT)))
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_timing(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        key = s.register(rd, selectors.EVENT_READ)
+
+        self.make_alarm(SHORT_SELECT, lambda *args: wr.send(b'x'))
+
+        with self.assertTakesTime(upper=SHORT_SELECT * 2):
```

Comment:
I think in each place we put tolerance that is *beyond* the tolerance that is baked into this helper, we should really put a comment explaining why we think we had to do this.